smithy-model/src/main/java/software/amazon/smithy/model/knowledge/PropertyBindingIndex.java

```diff
@@ -0,0 +1,206 @@
+
+                            .contains(memberShape.getMemberName())) {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(), true);
+                    } else {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(),
+                                doesNotRequireProperty(memberShape));
+                    }
+                    if (doesMemberShapeRequireProperty(memberShape.getId())
+                            || propertyNames.contains(memberShape.getMemberName())) {
+                        memberShapeToPropertyName.put(memberShape.getId(), getPropertyTraitName(memberShape)
+                            .orElse(memberShape.getMemberName()));
+                    }
+                }
+                // nesting is taking place, so index top level input/output members as not property.
+                if (!outputPropertiesShape.getId().equals(operationShape.getOutputShape())) {
+                    for (MemberShape memberShape : model.expectShape(operationShape.getOutputShape()).members()) {
+                        memberShapeDoesNotRequireProperty.put(memberShape.toShapeId(), true);
+                    }
+                }
+            }
+        }
+    }
+
+    public static PropertyBindingIndex of(Model model) {
+        return model.getKnowledge(PropertyBindingIndex.class, PropertyBindingIndex::new);
+    }
+
+    /**
+     * Gets the property name for a given member shape. Returns empty optional if the
+     * member shape does not correspond to a property.
+     *
+     * @param memberShapeId the ShapeId of the member shape to get the property name for.
+     * @return the property name for a given member shape if there is one.
+     */
+    public Optional<String> getPropertyName(ShapeId memberShapeId) {
+        return Optional.ofNullable(memberShapeToPropertyName.get(memberShapeId));
+    }
+
+    /**
+     * Returns true if a member shape positively maps to a property.
+     *
+     * {@see PropertyBindingIndex#getPropertyName(ShapeId)} will return a non-empty Optional if this method
+     * returns true.
+     *
+     * @param memberShapeId the ShapeId of the member shape to check
+     * @return true if member shape maps to a property on the given resource
+     */
+    public boolean isMemberShapeProperty(ShapeId memberShapeId) {
+        return memberShapeToPropertyName.containsKey(memberShapeId);
+    }
+
+    /**
+     * Resolves and returns the output shape of an operation that contains the
+     * top-level resource bound properties. Handles adjustments made with
+     * @nestedProperties trait.
+     *
+     * @param operation operation to retrieve output properties shape for.
+     * @return the output shape of an operation that contains top-level resource
+     *  properties.
+     */
+    public Shape getOutputPropertiesShape(OperationShape operation) {
```

Comment:
Wouldn't this always be a StructureShape? Would be nice to return it that way if so.

Reply:
`@nestedProperties` selector doesn't restrict what the target shape is so in theory here could return another shape type. However is all of those cases (will have to examine what behavior would happen for UnionShape) the abstract behavior is safe and correct.

Reply:
Hm, wouldn't nestedProperties only ever be able to target a structure? I don't see how it could work with any other shape, including a union. Seems like a restriction worth having, and then this API can be improved too.

Reply:
Adjusted.